modules/cuml/src/umap.ts

```diff
@@ -0,0 +1,377 @@
+   *   float32
+   * @param nFeatures number of features in the input X, if X is of the format [x1,y1,x2,y2...]
+   * @param returnType Desired output type of results and attributes of the estimators
+   *
+   * @returns Embedding of the data in low-dimensional space
+   */
+  fitTransformSeries<T extends Series<Numeric>, R extends Series<Numeric>, B extends boolean>(
+    X: T, y: null|R, convertDType: B, nFeatures = 1) {
+    this.fitSeries(X, y, convertDType, nFeatures);
+    const nSamples = Math.floor(X.length / nFeatures);
+    const dtype    = (convertDType) ? new Float32 : X.type as T['type'];
+
+    return this._process_embeddings(this._embeddings, nSamples, dtype) as
+           B extends true ? returnTypeMap<O, Float32>: returnTypeMap<O, T['type']>;
+  }
+
+  /**
+   * Fit X into an embedded space and return that transformed output.
+   *
+   *  There is a subtle difference between calling fit_transform(X) and calling fit().transform().
+   *  Calling fit_transform(X) will train the embeddings on X and return the embeddings. Calling
+   *  fit(X).transform(X) will train the embeddings on X and then run a second optimization.
+   *
+   *
+   * @param X Dense or sparse matrix containing floats or doubles. Acceptable dense formats: cuDF
+   *   DataFrame
+   * @param y cuDF Series containing target values
+   * @param convertDType When set to True, the method will automatically convert the inputs to
+   *   float32
+   * @param returnType Desired output type of results and attributes of the estimators
+   *
+   * @returns Embedding of the data in low-dimensional space
+   */
+  fitTransformDF<T extends Numeric, R extends Series<Numeric>, K extends string, B extends boolean>(
+    X: DataFrame<{[P in K]: T}>, y: null|R, convertDType: B) {
+    const nSamples = X.numRows;
+    this.fitDF(X, y, convertDType);
+    const dtype = X.get(X.names[0]).type as T;
+    return this._process_embeddings(this._embeddings, nSamples, dtype) as
+           B extends true ? returnTypeMap<O, Float32>: returnTypeMap<O, T>;
+  }
+
+  /**
+   * Fit X into an embedded space and return that transformed output.
+   *
+   *  There is a subtle difference between calling fit_transform(X) and calling fit().transform().
+   *  Calling fit_transform(X) will train the embeddings on X and return the embeddings. Calling
+   *  fit(X).transform(X) will train the embeddings on X and then run a second optimization.
+   *
+   *
+   * @param X array containing floats or doubles in the format [x1, y1, z1, x2, y2, z2...] for
+   *   features x, y & z.
+   * @param y array containing target values
+   * @param convertDType When set to True, the method will automatically convert the inputs to
+   *   float32
+   * @param nFeatures number of features in the input X, if X is of the format [x1,y1,x2,y2...]
+   * @param returnType Desired output type of results and attributes of the estimators
+   *
+   * @returns Embedding of the data in low-dimensional space
```

Comment:
Could use a little more detail about the return type, e,g. at least what keys will be in the return type map object

Reply:
updated